### PR TITLE
add missing threading methods

### PR DESCRIPTION
**Pull request by zsedem**
#9

### CODE REVIEW

File: lib/CPython.chs [hunk 1]
```diff
@@ -91,6 +99,7 @@ import           CPython.Internal
 {# fun Py_Finalize as finalize
   {} -> `()' id #}
 
+
```
Comment (mitchellvitez) [inline lib/CPython.chs:102]: random newline?

File: lib/CPython.chs [hunk 2]
```diff
@@ -164,6 +173,45 @@ endInterpreter :: ThreadState -> IO ()
...
+--
+-- Note: Calling this function from a thread when the runtime is finalizing
+-- will terminate the thread.
+gilStateEnsure :: IO GilState
+gilStateEnsure = do
+  state <- {# call PyGILState_Ensure as ^ #}
+  return $ GilState state
+
+-- | Release any resources previously acquired. After this call,
+-- Python’s -- state will be the same as it was prior to the corresponding
```
Comment (mitchellvitez) [inline lib/CPython.chs:194]:
```suggestion
-- Python’s state will be the same as it was prior to the corresponding
```

File: lib/CPython.chs [hunk 3]
```diff
@@ -164,6 +173,45 @@ endInterpreter :: ThreadState -> IO ()
...
+-- regardless of the current state of Python, or of the global interpreter
+-- lock. This method is generally safest way to make python calls in
+-- a multi-threaded scenario, however if you calling this from not bound
+-- threads, it may have some performance implications.
+withGilStateEnsure :: IO a -> IO a
+withGilStateEnsure ioOperation = runInBoundThread $ do
+  gstate <- gilStateEnsure
+  result <- ioOperation
+  gilStateRelease gstate
+  return result
```
Comment (mitchellvitez) [inline lib/CPython.chs:213]: Might want to use [`bracket`](https://hackage.haskell.org/package/base-4.14.1.0/docs/src/Control.Exception.Base.html#bracket) here

**Review by mitchellvitez — APPROVED** [review]
Couple small suggestions, otherwise looks like a reasonable wrapper over the C calls

File: lib/CPython.chs (hunk 3 above)
Comment (zsedem) [reply lib/CPython.chs:213]: :+1:

**Comment by zsedem** [comment]
My conclusion on this is that these methods are just not enough. Also I confirmed, that if you compile your project with `-threaded` it will definitely throw segmentation fault eventually.

So if we want to support the `-threaded` mode, then we need to use `PyEval_SaveThread` and `PyEval_RestoreThread` methods, which sounds easy (at first).

The problem is that we need to call it even at the bottom of the callstack too, where we use the `incref` and `decref` methods for objects created/owned by Haskell. So these calls at the bottom of the callstack need the `PyThreadState` to acquire GIL. For this problem I created a global variable, where the threadstate can rest until the next call will need it to acquire the GIL (quite messy, but works).

This introduces a new problem :smile: : If you start using the subinterpreters, then the methods you are calling should acquire the GIL for the subinterpreter for that thread :man_facepalming: .

Anyway I think the best solution would be to create some Monad for python codes, which can store these states:
* subinterpreter for this thread
* ThreadState for acquiring and releasing the GIL
This would ensure that all Python API calls are in this monad, we could release the GIL, when code is running, which is normal IO monad.

**Comment by mitchellvitez** [comment]
One other small-but-nice thing about a `Python` monad would be users wouldn't have to remember to call `initialize` etc.

**Pull request closed without merging**